File: Frontend/src/services/amenity.service.ts
import axiosInstance from "./core/axiosinstance";

export interface Amenity {
    id: number;
    name: string;
    description: string;
    rules: string;
    is_active: boolean;
}

export interface AmenityBooking {
    id: number;
    amenity: number;
    start_time: string;
    end_time: string;
    status: 'Pending' | 'Approved' | 'Rejected' | 'Cancelled';
    purpose: string;
}

export const fetchAmenities = async () => {
    const response = await axiosInstance.get('/amenities/');
    return response.data;
}

export const fetchAmenityBookings = async () => {
    const response = await axiosInstance.get('/amenities/bookings/');
    return response.data;
}

export const createBooking = async (data: { amenity: number; start_time: string; end_time: string; purpose: string }) => {
    const response = await axiosInstance.post('/amenities/bookings/', data);
    return response.data;
}

export const approveBooking = async (bookingId: number) => {
    const response = await axiosInstance.post(`/amenities/bookings/${bookingId}/approve/`);
    return response.data;
}

export const rejectBooking = async (bookingId: number) => {
    const response = await axiosInstance.post(`/amenities/bookings/${bookingId}/reject/`);
    return response.data;
}

export const cancelBooking = async (bookingId: number) => {
    const response = await axiosInstance.post(`/amenities/bookings/${bookingId}/cancel/`);
    return response.data;
}

File: Backend/apt_proj/Apt_Amenities/services/amenity_service.py
from django.core.exceptions import ValidationError
from django.db.models import Q
from ..Amenities_models import AmenityBooking
from apt_proj.Apt_Notifications.tasks import send_notification_task

def create_booking(user, amenity, start_time, end_time, purpose):
    # prevent creating a booking if the timeslot overlaps with an APPROVED/CONFIRMED booking
    overlapping = AmenityBooking.objects.filter(
        amenity=amenity,
        status=AmenityBooking.Status.CONFIRMED,
        start_time__lt=end_time,
        end_time__gt=start_time
    ).exists()

    if overlapping:
        raise ValidationError("The selected timeslot overlaps with an existing booking")
    booking = AmenityBooking.objects.create(
        user=user,
        amenity=amenity,
        start_time=start_time,
        end_time=end_time,
        purpose=purpose,
        status = AmenityBooking.Status.PENDING
    )
    return booking

def update_booking_status(booking, new_status, manager):
    if booking.status == new_status: return booking

    # check overlap before approving - catches race conditions for 2 pending bookings

    if new_status == AmenityBooking.Status.CONFIRMED:
        overlapping = AmenityBooking.objects.filter(
            amenity=booking.amenity,
            status=AmenityBooking.Status.CONFIRMED,
            start_time__lt=booking.end_time,
            end_time__gt=booking.start_time
        ).exclude(id=booking.id).exists()

        if overlapping:
            raise ValidationError("Overlapping booking exists - cannot approve")

    booking.status = new_status
    booking.save(update_fields = ['status', 'updated_at'])

    # send notification
    if new_status in [AmenityBooking.Status.CONFIRMED, AmenityBooking.Status.REJECTED]:
        send_notification_task.delay(
            title=f"Amenity Booking {new_status.title()}",
            body=f"Your booking for {booking.amenity.name} has been {new_status.lower()}.",
            data={"event_type": "booking_updated", "booking_id": booking.id},
            user_ids=[booking.user.id],
            all_users=False
        )
    return booking

def cancel_booking(booking, user):
    if booking.user != user and not(hasattr(user, 'profile') and user.profile.role.name in ['admin', 'manager']):
        raise ValidationError("You can only cancel your own bookings")
    
    if booking.status in [AmenityBooking.Status.REJECTED, AmenityBooking.Status.CANCELLED]:
        raise ValidationError("Booking is already rejected or cancelled.")

    booking.status = AmenityBooking.Status.CANCELLED
    booking.save(update_fields = ['status', 'updated_at'])
    return booking